lib/pbench/common/utils.py

```diff
@@ -62,3 +74,120 @@ def validate_hostname(host_name: str) -> int:
         return 1
 
     return 0
+
+
+class CleanupNotCallable(Exception):
+    """
+    Signal that caller tried to register a cleanup action with an object that
+    was not a Callable on an object.
+    """
+
+    def __init__(self, action):
+        self.action = action
+
+    def __str__(self) -> str:
+        return f"Parameter {self.action!r} ({type(self.action)} is not a Callable"
+
+
+# Following are a set of classes to perform hierarchical cleanup actions when
+# an error occurs.
+#
+# A Deque supports an ordered list of cleanup actions that will be popped and
+# executed in reverse order on demand.
+#
+# Cleanup actions are Callable objects; to queue an action requiring parameters
+# a `lambda` expression can be used.
+
+
+class CleanupAction:
+    """
+    Define a single cleanup action necessary to reverse persistent steps in the
+    upload procedure, based on the Step ENUM associated with the action.
+    """
+
+    def __init__(self, logger: Logger, action: Callable, name: str = None):
+        """
+        Define a cleanup action
+
+        Args:
+            logger: The active Pbench Logger object
+            action: a Callable to perform cleanup
+            name: optional printable name for debugging
+        """
+        self.action = action
+        self.logger = logger
+        self.name = name if name else repr(action)
+
+    def cleanup(self):
+        """
+        Perform a cleanup action, executing a callable associated with some
+        object (usually a Dataset or a Path) that needs cleaning.
+
+        This handles errors and reports them, but doesn't propagate failure to
+        ensure that cleanup continues as best we can.
+        """
+        try:
+            self.action()
+        except Exception as e:
+            # TODO: f-string used because this is shared by agent and server
+            self.logger.exception(f"Unable to {self}: {e}")
```

Comment:
Well, we have a problem.  The logging on the agent side still uses `%s` style format strings.

Reply:
Ha!  Credit goes to Dave for hanging a lantern on it.  😁